# Identifying-phages-from-metagenomes



"What about the phages?" is a question I often see raised by microbiologists. It's easy to neglect the phage fraction and all too common to forget about these microbial parasites. Phages are uniquely difficult to study due to their profound diversity and lack of a single marker gene that complicates sequencing. The result is that the vast majority of phages are unidentified. 

To overcome the obstacle of no universal phage gene, metagenomic sequencing is used to capture phage DNA from samples. The high-throughput output of metagenomics is fascinating, but this immense amount of data also poses a complicated barrier to entry.
(continue here talking about this is why wet lab microbiologists dont look at phages). 


Metagenomics can be daunting for microbiologists to start exploring, especially for those with little to no experience in bioinformatics. The high-throughput of metagenomics can be intimidating and I often from fellow microbiologists that they do not exactly know where to begin. 16S and amplicon sequencing is significantly easier 

Numerous times, I have encountered fellow microbiologists who have had limited/no experience with bioinformatics

This is a walkthrough on how to recover phage and prophage genomes from short or long read metagenomic datasets. 
